### DISCUSSION TITLE
Custom state backends

### DISCUSSION THREAD

**Discussion by mjsqu** [meltano/meltano#8276] [Ideas]
I managed to stand up a [working feature branch of meltano](https://github.com/mjsqu/meltano/tree/feat/aws_ssm_state_backend) using AWS SSM Parameter Store as its state backend. The following state commands were analogous to the boto3 ssm client functions:

|meltano state ... command|ssm client function|
|---|---|
|list|`get_parameters_by_path(Path=)`|
|set|`put_parameter(Name=,Value=)`|
|get|`get_parameter(Name=)`|
|clear|`delete_parameter(Name=)`|

The coding I had to do was quite minimal as the `meltano/src/core/state_store` modules are really well written and provide a lot of the functionality already. The SSM module I added was a mash-up of the `s3.py` and the `db.py` modules:

- s3.py - contributed logic for connection to aws via boto3
- db.py - contributed the get() and set() methods amongst others, and meant I didn't need to declare my backend as a 'filesystem' type

It would be interesting to explore how meltano could provide a method of registering state backends in the project in the same way that taps/targets can be registered, should users wish to store their state somewhere different, such as:

- AWS Systems Manager Parameter Store (SSM)
- Google Cloud DataStore
- Azure App Configuration
- Any NoSQL service that provides key-value pair storage
- A REST API which exposes GET/PUT/DELETE endpoints for each meltano state operation

**Comment by mjsqu** [comment]
From: https://meltano.slack.com/archives/C013Z450LCD/p1700177112720099?thread_ts=1700118139.892419&cid=C013Z450LCD

(https://discuss.meltano.com/c/contributing - when the above Slack message expires)

**Reply by edgarrmondragon** [reply]
https://github.com/meltano/meltano/tree/state-backend-plugins 🙂

**Reply by mjsqu** [reply]
Man it is hard work trying to google "python plugin" and get some sort of guide to developing what I'm wanting, rather than a plugin for all sorts of software!

Anyway, to my surprise, I managed to splice the feature branch I'd added to my fork of meltano into my own package:

https://github.com/mjsqu/meltano-state-ssm-plugin

Salient part of the `pyproject.toml` being:

```toml
[tool.poetry.plugins."meltano.state_backends"]
"ssm-parameter" = "meltano_state_ssm_plugin.ssm:SSMStateStoreManager"
```

Which I then ran with your branch, 

```
meltano init meltano-state-ssm
cd meltano-state-ssm/

#  Verify we can run meltano state as usual
meltano state
meltano state list

# Set the ssm-parameter backend
meltano config meltano set state_backend.uri "ssm-parameter://meltano/state"
meltano state list
```

Jump into python and import `StateBackend` from `meltano.core.state_store`:

```python
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from meltano.core.state_store import StateBackend
>>> StateBackend.backends
<bound method StateBackend.backends of <class 'meltano.core.state_store.StateBackend'>>
>>> list(StateBackend.backends)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'method' object is not iterable
>>> list(StateBackend.backends())
[<BuiltinStateBackendEnum.SYSTEMDB: 'systemdb'>, <BuiltinStateBackendEnum.LOCAL_FILESYSTEM: 'file'>, <BuiltinStateBackendEnum.AZURE: 'azure'>, <BuiltinStateBackendEnum.S3: 's3'>, <BuiltinStateBackendEnum.GCS: 'gs'>, 'meltano-state-ssm-plugin']
>>> quit()
```

yay 🚀  - meltano-state-ssm-plugin is listed

Oops, it should be called ssm-parameter in my plugin's `pyproject.toml` file:

![image](https://github.com/meltano/meltano/assets/17741131/46352fe1-5b92-4e18-8532-857af71a234e)

And finally after some more messing around with the path to the module:

![image](https://github.com/meltano/meltano/assets/17741131/1186e82e-6d7a-44a2-84b2-a65e7a0967dc)

The message "You must specify a region" confirms boto3 is having a go at getting state.

**Reply by mjsqu** [reply]
And finally, providing some credentials, I was able to demonstrate working `list`, `set` and `get` commands:
![image](https://github.com/meltano/meltano/assets/17741131/9755dcd2-ce75-4f3e-9021-7675c582e3a1)

Thanks for the help!

I'd really appreciate it if you could review my `pyproject.toml` as I've never written a Python package before, let alone a plugin - and also it would be good to get some guidance on naming conventions for the package, as I will probably try and work on Google Cloud DataStore and Azure next and want to lock in a solid naming standard, ideas:

meltano-state-aws-ssm-<backend|plugin|pluggable>
meltano-state-gcs-datastore-<backend|plugin|pluggable>

**Reply by edgarrmondragon** [reply]
Hey @mjsqu! I'll give your `pyproject.toml` a look but I actually similarly created a sample "plugin" package, though it uses Hatch instead of Poetry: https://github.com/edgarrmondragon/meltano-rocksdb-state-backend

**Reply by mjsqu** [reply]
Even better, I'll look over your repo as well! Thank you.

**Reply by mjsqu** [reply]
`meltano-<service>-state-backend` is a good naming convention, I'll rename mine

**Reply by mjsqu** [reply]
I imagine this could live elsewhere in future, but I made a very early start on a cookiecutter template for state-backends:
https://github.com/mjsqu/cookiecutter-meltano-state-backend

**Comment by edgarrmondragon** [comment]
These will be shipped in the next release!
